passes/techmap/iopadmap.cc

```diff
@@ -192,53 +192,97 @@ struct IopadmapPass : public Pass {
 			if (!toutpad_celltype.empty() || !tinoutpad_celltype.empty())
 			{
 				dict<SigBit, Cell *> tbuf_bits;
+				pool<SigBit> driven_bits;
 
+				// Gather tristate buffers and always-on drivers.
 				for (auto cell : module->cells())
 					if (cell->type == ID($_TBUF_)) {
 						SigBit bit = cell->getPort(ID::Y).as_bit();
 						tbuf_bits[bit] = cell;
+					} else {
+						for (auto port : cell->connections())
+							if (!cell->known() || cell->output(port.first))
+								for (auto bit : port.second)
+									driven_bits.insert(bit);
+					}
+
+				// If a wire is a target of an assignment, it is driven, unless the source is 'z.
+				for (auto &conn : module->connections())
+					for (int i = 0; i < GetSize(conn.first); i++) {
+						SigBit dstbit = conn.first[i];
+						SigBit srcbit = conn.second[i];
+						if (!srcbit.wire && srcbit.data == State::Sz)
+							continue;
+						driven_bits.insert(dstbit);
 					}
 
 				for (auto wire : module->selected_wires())
 				{
 					if (!wire->port_output)
 						continue;
 
+					// Don't handle inout ports if we have no suitable buffer type.
```

Comment:
These feel like they should emit warnings to the user if they ever trigger.

Reply:
Not necessarily — in theory someone could be running `iopadmap` in several passes to insert various types of buffers (though I can't think of a combination where you'd have tinoutpad but not toutpad or the other way around — mostly you'd be inserting tristate buffers separately from non-tristate ones).  @eddiehung what do you think?

Reply:
Hmm. I think if the user only specifies `-tinoutpad` or `-toutpad` but not the other then they would likely have a good reason for doing so; failing that a `$_TBUF_` would be left behind anyway. Perhaps a debug message might be useful here just in case?

Reply:
I decided against adding it, it would require a refactor to do it right — if we just put a warning here, it will trigger whether tristate buffers are actually present or not,